invokeai/app/invocations/baseinvocation.py

```diff
@@ -3,15 +3,309 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
+from enum import Enum
 from inspect import signature
-from typing import TYPE_CHECKING, Dict, List, Literal, TypedDict, get_args, get_type_hints
+from typing import TYPE_CHECKING, AbstractSet, Any, Mapping, Optional, Type, Union, get_args, get_type_hints
 
-from pydantic import BaseConfig, BaseModel, Field
+from pydantic import BaseModel, Field
+from pydantic.fields import Undefined
+from pydantic.typing import NoArgAnyCallable
 
 if TYPE_CHECKING:
     from ..services.invocation_services import InvocationServices
 
 
+class Input(str, Enum):
+    """
+    The type of input a field accepts.
+    - `Input.Direct`: The field must have its value provided directly, when the invocation and field \
+      are instantiated.
+    - `Input.Connection`: The field must have its value provided by a connection.
+    - `Input.Any`: The field may have its value provided either directly or by a connection.
+    """
+
+    Connection = "connection"
+    Direct = "direct"
+    Any = "any"
+
+
+class UITypeHint(str, Enum):
+    """
+    Type hints for the UI.
+    If a field should be provided a data type that does not exactly match the python type of the field, \
+    use this to provide the type that should be used instead.
```

Comment:
Adding a new type to the UI is rather cumbersome - you need to edit 4 or 5 files, including TS types, runtime logic, and UI components. I'd rather include some documentation in the node development docs (currently WIP), rather than in the code.

Reply:
Added a brief statement to look for the node dev docs